#!/bin/bash
#PBS -j oe
#PBS -m ae
#PBS -N 10_finalcall
#PBS -l select=1:ncpus=5:mem=10gb
#PBS -l walltime=5:00:00
#PBS -M your@email.au

ncpu=$(qstat -f $PBS_JOBID | grep "Resource_List.ncpus" | cut -d= -f2 | sed 's/ //g') 
s=$(if [ $ncpu == 1 ]; then echo ""; else echo "s"; fi)
mem=$(qstat -f $PBS_JOBID | grep "Resource_List.mem" | cut -d= -f2 | sed 's/ //g')
echo $(date)
echo "------------------------------------------------------"
echo "This job is allocated "$ncpu" CPU core$s and $mem on " cat $PBS_NODEFILE | uniq
echo "------------------------------------------------------"
echo "Submitted to $PBS_QUEUE@$PBS_O_HOST"
echo "Working directory is $PBS_O_WORKDIR"
echo "Job identifier is $PBS_JOBID"
echo "Job name is $PBS_JOBNAME"
echo "------------------------------------------------------"

cd path/to/snp_calling/
shopt -s expand_aliases
source /etc/profile.d/modules.sh

# Load modules
module load conda3
source $CONDA_PROF/conda.sh
conda activate gatk4-4.2.0.0

# The ocd box
REFERENCE="$PWD/reference/reference.unaligned.fasta"
COMBINED_GVCF="$PWD/10_haplotype_cohort_output/combined_gvcf_recal2.vcf"
OUTDIR1="$PWD/10a_haplotype_cohort_output"
OUTDIR2="$PWD/10b_GVCF_output"
OUTDIR3="$PWD/10c_hardfiltering_output"
LOGDIR1="$PWD/logs/10a_haplotype_cohort_logs"
LOGDIR2="$PWD/logs/10bc_GVCF_hardfiltering_logs"

############################ HAPLOTYPE CALLER ARRAY ############################## Comment-out after running 
# # Path to recalibrated bam files
# RECAL_BAMS=($(ls $PWD/9_BQSR_recal2_output/*recal2.bams))
# # Define array
# id=${PBS_ARRAY_INDEX}
# SAMPLE=${RECAL_BAMS[$id]}

# # Taxon or sample we are working now
# echo "Processing $SAMPLE"
# # Get sample names
# OUTPUT_BASENAME=$(echo $SAMPLE | cut -d/ -f7)
# echo $OUTPUT_BASENAME
# OUTPUT_NAME=$(echo $OUTPUT_BASENAME | cut -d. -f1).g.vcf
# echo $OUTPUT_NAME

# # Final run of variant calling on recalibrated bam files
# gatk HaplotypeCaller \
#     -R ${REFERENCE} \
#     -I ${SAMPLE} \
#     -O ${OUTDIR1}/${OUTPUT_NAME} \
#     --emit-ref-confidence GVCF \
#     --contamination-fraction-to-filter 0.0002 \
#     --min-base-quality-score 20 \
#     --phred-scaled-global-read-mismapping-rate 30 \
#     --standard-min-confidence-threshold-for-calling 40.0 &

# wait

# # clean logs away once done running - move the last one manually
# mv $PBS_O_WORKDIR/$PBS_JOBNAME.* $LOGDIR1

########################################################################################
# When you've ran your last haplotype caller, remove #PBS -J 0-90 to deactivate the array job and comment-out last step

# Consolidate/Combine GVCFs files to create the file for joint genotyping (t=15mins)
ls -d -1 ${OUTDIR1}/*.g.vcf > ${OUTDIR1}/gvcf_recal2.list &

wait

# Consolidate gvcf files (t=13mins)
gatk CombineGVCFs \
    -R ${REFERENCE} \
    -V ${OUTDIR1}/gvcf_recal2.list \
    -O ${OUTDIR1}/combined_gvcf_recal2.vcf &
#--dbsnp $DBSNP

wait

## JOINT GENOTYPING ##
gatk GenotypeGVCFs \
    -R ${REFERENCE} \
    -V ${COMBINED_GVCF} \
    -O ${OUTDIR2}/genotyped_X_samples_recal2.vcf &

wait

#Extract SNPSs from callset
gatk SelectVariants \
	-R ${REFERENCE} \
	-V ${OUTDIR2}/genotyped_X_samples_recal2.vcf \
	-select-type SNP \
	-O ${OUTDIR2}/genotyped_X_samples_snps_recal2.vcf &&

# Extract indels from callset
gatk SelectVariants \
	-R ${REFERENCE} \
	-V ${OUTDIR2}/genotyped_X_samples_recal2.vcf \
	-select-type INDEL \
	-O ${OUTDIR2}/genotyped_X_samples_indels_recal2.vcf &

wait

## HARDFILTERING ##

gatk VariantFiltration \
    -R ${REFERENCE}  \
    -V ${OUTDIR2}/genotyped_X_samples_snps_recal2.vcf \
    -O ${OUTDIR3}/genotyped_X_samples_snps_filtered_recal2.vcf \
    --cluster-window-size 10 \
    --filter-expression "QUAL < 30.0" \
    --filter-name "LowQuality" \
    --filter-expression "QD < 4.0" \
    --filter-name "QualByDepth" \
    --filter-expression "MQ < 38.0" \
    --filter-name "RMSMappingQuality" \
    --filter-expression "FS > 60.0" \
    --filter-name "FisherStrand" &&

# Filter variants according to filter expression: INDEL
gatk VariantFiltration \
	-R ${REFERENCE} \
	-V ${OUTDIR2}/genotyped_X_samples_indels_recal2.vcf \
  -O ${OUTDIR3}/genotyped_X_samples_indels_filtered_recal2.vcf \
	-filter-name "QD_filter" \
  -filter-expression "QD < 2.0" \
	-filter-name "FisherStrand_filter" \
  -filter-expression "FS > 200.0" \
	-filter-name "LowQual_filter" \
  -filter-expression "QUAL < 30.0" &

wait

# Exclude filtered variants to get only the variants that PASSED the filtering as the input for the BaseRecalibration (BQSR) tool later on because, like many GATK tools, it might not ignore the "FAILED" variants status so better be on the safe side.

gatk SelectVariants \
    --exclude-filtered true \
    -V ${OUTDIR3}/genotyped_X_samples_snps_filtered_recal2.vcf \
    -O ${OUTDIR3}/PMCO_genotyped_PASS_snps_recal2.vcf &&

gatk SelectVariants \
    --exclude-filtered true \
    -V ${OUTDIR3}/genotyped_X_samples_indels_filtered_recal2.vcf \
    -O ${OUTDIR3}/PMCO_genotyped_PASS_indels_recal2.vcf &

wait

# clean logs away once done running - move the last one manually
mv $PBS_O_WORKDIR/$PBS_JOBNAME.* ${LOGDIR2}

